### PR TITLE
fix: Upgraded dependencies in Cargo.toml

### PR DESCRIPTION
**Pull request by SiddheshKanawade**
I have Updated the dependencies to their latest version. I have also done basic migration towards Rust 2021 version from 2018 version

### CODE REVIEW

**Comment by hsluoyz** [comment]
@PsiACE

**Comment by PsiACE** [comment]
hi @SiddheshKanawade , Thank you very much for the PR, I need some time to check it. 

Can you try adding some github actions to help check the format and clippy?

**Comment by SiddheshKanawade** [comment]
> hi @SiddheshKanawade , Thank you very much for the PR, I need some time to check it.
> 
> Can you try adding some github actions to help check the format and clippy?

@PsiACE sure, I will try to integrate some GitHub actions

**Comment by SiddheshKanawade** [comment]
@PsiACE why do we only make GitHub Actions for ubuntu? shall I include macOS and windows too using `${{ matrix.os }}`

**Comment by SiddheshKanawade** [comment]
> hi @SiddheshKanawade , Thank you very much for the PR, I need some time to check it.
> 
> Can you try adding some github actions to help check the format and clippy?

Also, I could see a file named `check.yml` this has the checks for formatting and clippy, do you wish to upgrade the existing file or make another separately for formatting and check?

**Comment by PsiACE** [comment]
> shall I include macOS and windows too using ${{ matrix.os }}

Yes, please

>  do you wish to upgrade the existing file or make another separately for formatting and check?

Just upgrade it .

**Comment by SiddheshKanawade** [comment]
@PsiACE can you please review the file at [check.yml](https://github.com/SiddheshKanawade/Rust_Tutorial/blob/master/.github/workflows/check_clippy.yml)

I have added some CI tests and rest are formatting and clippy checks, can you please review it and let me know its status. If is bids the correctness I will commit it to this repo

**Comment by SiddheshKanawade** [comment]
@PsiACE I just noticed that in the file at check.yml, I am having some syntax errors, kindly ignore them for now and check the rest, I will correct them meanwhile

**Comment by PsiACE** [comment]
https://github.com/SiddheshKanawade/Rust_Tutorial/blob/6138066d0a7ff800032125e84b7eeed401e94129/.github/workflows/check_clippy.yml#L4

good

https://github.com/SiddheshKanawade/Rust_Tutorial/blob/6138066d0a7ff800032125e84b7eeed401e94129/.github/workflows/check_clippy.yml#L33

we don't need wasm target for examples

https://github.com/SiddheshKanawade/Rust_Tutorial/blob/6138066d0a7ff800032125e84b7eeed401e94129/.github/workflows/check_clippy.yml#L52-L68

consider use a easy cache https://github.com/Swatinem/rust-cache

**Comment by PsiACE** [comment]
Also, since this repository is some examples, we don't need to check all kinds of features, just build, test and clippy.

**Comment by SiddheshKanawade** [comment]
Sure I will work on it

**Comment by PsiACE** [comment]
Another fact is that the current [check.yml](https://github.com/casbin-rs/examples/blob/master/.github/workflows/check.yml) is available, all you need to do is add the cache and allow it to check PR.

**Comment by SiddheshKanawade** [comment]
@PsiACE can you review it once again at [check.yml](https://github.com/SiddheshKanawade/Rust_Tutorial/blob/master/.github/workflows/check_clippy.yml)

I have implemented only basic checks on `build`, `clippy`, `format` and `cache`. Checks for `wasm`, `Tokio`, `async-std` have been removed since this is an example repository. I have also not involved command to install `gnu-tar` which is specific for Mac-OS since I felt it just allows to create tar files and that may not be necessary for this repository.

Kindly review the file and let me know whether should I commit it to this repository or some more changes are needed

**Comment by SiddheshKanawade** [comment]
> @PsiACE can you review it once again at [check.yml](https://github.com/SiddheshKanawade/Rust_Tutorial/blob/master/.github/workflows/check_clippy.yml)
> 
> I have implemented only basic checks on `build`, `clippy`, `format` and `cache`. Checks for `wasm`, `Tokio`, `async-std` have been removed since this is an example repository. I have also not involved command to install `gnu-tar` which is specific for Mac-OS since I felt it just allows to create tar files and that may not be necessary for this repository.
> 
> Kindly review the file and let me know whether should I commit it to this repository or some more changes are needed

@PsiACE is this fine or any changes are required?

**Comment by PsiACE** [comment]
> @PsiACE is this fine or any changes are required?

`Generate test secret key` is useful, we need to keep it.

also , I'm not sure if `check cache` will work. 

I think you can refer to https://github.com/engula/engula/blob/5d8e4b2362582bf5e475d032a0704b745f9d4106/.github/workflows/ci.yml#L136

**Comment by PsiACE** [comment]
@SiddheshKanawade I suggest you commit it to this pr, it helps to check if there is a problem.

**Comment by SiddheshKanawade** [comment]
> > @PsiACE is this fine or any changes are required?
> 
> `Generate test secret key` is useful, we need to keep it.
> 
> also , I'm not sure if `check cache` will work.
> 
> I think you can refer to https://github.com/engula/engula/blob/5d8e4b2362582bf5e475d032a0704b745f9d4106/.github/workflows/ci.yml#L136

yes, `Generate test key` will be there, its just that it was another repository and a error would have been prompted hence I commented it out.

Ok, I will commit it to this PR

**Comment by SiddheshKanawade** [comment]
@PsiACE can you please review, I have committed the changes

**Comment by SiddheshKanawade** [comment]
@PsiACE do I need to upgrade cache_check?

**Comment by PsiACE** [comment]
> @PsiACE do I need to upgrade cache_check?

Of course, it doesn't work now.

Also, try to make the ci pass.

**Comment by SiddheshKanawade** [comment]
@PsiACE can you please review

**Comment by hsluoyz** [comment]
@SiddheshKanawade plz fix:

image

**Comment by SiddheshKanawade** [comment]
> @SiddheshKanawade plz fix:
> 
> image

Yes, this is the next task in list, I will try to ensure that the repository passes CI test.

@PsiACE, I could see that `swatinem rust/cache` is yet not working.

I am not able to go to the root cause of the problem. Till now I have tried to rule out the following possibilities which would have caused the problem
1. I have ensured that I call `rust-toolchain` before execution of cache, since `swatinem/rust-cache` may use the rust/rustc version as key.
2. Since `Swatinem/rust-cache` works well with projects that have `stable` toolchain running and has `Cargo.lock` in project, so I ensured that they are present in project by making `stable` as default toolchain and explicitly generating the `Cargo.lock` file.
3. Also, I tried using `actions/cache@v1` instead of `swatinem/rust-cache` with `restored-keys` to ensure that their isn't any error in mapping with key.
4. I also tried switching the operating system from `macOS` to `Ubuntu` as a rare possibility that local environment might have caused the error, but I was wrong there.
5. I have also tried to fork `encula/encula` and added on comment in file under `src` and committed the change, but their original code using `swatinem/rust-cache` didn't generate any `cache`.
6. I have also tried to cache only `./target` by explicitly writing it in `target-dir` even though it is by default `./target`.

I am not sure whether the problem is in code, or something else is causing the problem.

@PsiACE can you please guide me on this

**Comment by PsiACE** [comment]
hi @SiddheshKanawade If you want rust-cache to work, first fix the errors reported in the code. The cache can only complete the post if the ci can pass.

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -50,10 +71,14 @@ jobs:
 
   check_clippy:
     name: "Check Clippy"
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
```
Comment (PsiACE) [inline .github/workflows/check.yml]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -27,17 +42,23 @@ jobs:
...
+        cargo build --verbose
         cargo test --all --no-fail-fast
 
   check_formatting:
     name: "Check Formatting"
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
```
Comment (PsiACE) [inline .github/workflows/check.yml]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/check.yml [hunk 3]
```diff
@@ -11,11 +12,25 @@ on:
 jobs:
   test:
     name: "Test"
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
```
Comment (PsiACE) [inline .github/workflows/check.yml]:
```suggestion
        os: [ubuntu-latest, macOS-latest]
```

**Review by PsiACE — COMMENTED** [review]
It is enough to run clippy and fmt on ubuntu, which saves ci resources.

Windows is not needed because we may not have enough time to maintain it.

**Comment by SiddheshKanawade** [comment]
> hi @SiddheshKanawade If you want rust-cache to work, first fix the errors reported in the code. The cache can only complete the post if the ci can pass.

Ok, will try to resolve possible errors, thank you for the suggestion

**Comment by hsluoyz** [comment]
@PsiACE

**Comment by PsiACE** [comment]
> @PsiACE

The ci didn't trigger due to github's glitch yesterday, so I guess we can check it when the next commit is made.

**Comment by SiddheshKanawade** [comment]
@PsiACE sorry for the delayed reply but I had my mid semester examination going on. I could see that I have to change `default_model.rs`, `config.rs`, `builder.rs`, `test.rs` `file_adapter.rs` which are in the `Casbin-rs` repository to make `cargo build --verbose` to compile without failure. I am planning to use `cfg-if, cfg-else` instead of `#cfg` in `file_adapter.rs` so as to avoid repeated declarations and imports wherever possible. 

I will try to make new commit sooner so that we can check the CI.

**Pull request closed without merging**